Feature/categories feature domain and data and ui

DinaOthman21 opened the pull request:
add categories feature domain and data and ui
and update media exceptions

DinaOthman21 committed:
add movies Dto

DinaOthman21 committed:
add getMoviesByGenre in MovieApiService

DinaOthman21 committed:
update MovieRemoteDataSource and MovieRemoteDataSourceImpl

DinaOthman21 committed:
replace Dtos

DinaOthman21 committed:
replace getMoviesByCategory

DinaOthman21 committed:
update MovieRemoteDataSource

DinaOthman21 committed:
update MediaRemoteDataSource & MediaRemoteDataSourceImpl

DinaOthman21 committed:
update AflamiException

DinaOthman21 committed:
update Mapper

DinaOthman21 committed:
add getMoviesByCategory

DinaOthman21 committed:
Merge remote-tracking branch 'origin/develop' into feature/categories

DinaOthman21 committed:
add TvShowByCategoryDto

DinaOthman21 committed:
update MediaApiService

DinaOthman21 committed:
update MediaRemoteDataSource & MediaRemoteDataSourceImpl

DinaOthman21 committed:
update AflamiException

DinaOthman21 committed:
update MediaMapper

DinaOthman21 committed:
add getTvShowsByCategory

DinaOthman21 committed:
add GetMoviesByCategoryUseCase

DinaOthman21 committed:
add GetTvShowsByCategoryUseCase

DinaOthman21 committed:
inject use cases

DinaOthman21 committed:
add GetMoviesByCategoryUseCaseTest

DinaOthman21 committed:
add GetTvShowsByCategoryUseCaseTest

DinaOthman21 committed:
setup navigation

DinaOthman21 committed:
add text to CategoriesScreen

Joseph-Sameh-0 committed:
Merge branch 'develop' into feature/categories

Joseph-Sameh-0 committed:
Bump version to 0.8.2

Joseph-Sameh-0 committed:
Refactor: Rename CategoriesGraph1 to MainGraph and update state emission

- Renamed `CategoriesGraph1` to `MainGraph` in `CategoriesDestinations` for clarity.
- Updated references to `CategoriesGraph1` in `CategoriesNavGraph.kt` and `FeatureAPIModule.kt`.
- In `BaseViewModel.kt`:
    - Renamed `privateScreenState` to `_screenState`.
    - Renamed `emitState` function to `updateState`.

Joseph-Sameh-0 committed:
Refactor: Update CategoriesScreenViewModel and BaseViewModel

- Updated `CategoriesScreenViewModel` to simplify constructor and remove redundant initialization of `CategoriesScreenUIState`.
- Modified `BaseViewModel`'s `tryToExecute` function to pass the entire `Throwable` object to the `onError` lambda, instead of just the message string.
- Adjusted `CategoriesScreenUIState` and `CategoriesUIState` to provide default values for their properties.

Joseph-Sameh-0 committed:
Refactor: Restructure Categories feature and add CategoryDetailsScreen

This commit refactors the Categories feature by:
- Deleting `CategoriesDestination.kt` and `CategoriesGraph.kt`.
- Restructuring `CategoriesDestinations.kt` to include `CategoriesGraph` and a new `CategoryDetailsScreen` destination.
- Adding `CategoryDetailsScreen.kt`, `CategoryDetailsScreenViewModel.kt`, `CategoryDetailsScreenUIState.kt`, and `CategoryDetailsScreenInteractionListener.kt` for the new category details screen.
- Updating `CategoriesNavGraph.kt` to include the route for `CategoryDetailsScreen`.
- Renaming and moving `CategoriesScreen.kt`, `CategoriesScreenViewModel.kt`, `CategoriesScreenUIState.kt`, and `CategoriesScreenInteractionListener.kt` into a new `categories` subpackage.
- Adding previews for `CategoriesScreen` and `CategoryDetailsScreen`.

Joseph-Sameh-0 committed:
Refactor: Make CategoriesGraph a sealed interface and add dependencies

Updated `CategoriesGraph` to be a `sealed interface` for improved type safety in categories navigation.

Added Hilt and Kotlin Serialization dependencies to the `categoriesUi` module.

Joseph-Sameh-0 committed:
Refactor: Categories feature UI and state management

This commit introduces significant changes to the categories feature:

- Deleted `CategoryDetailsScreenInteractionListener`, `CategoryDetailsScreenUIState`, and `CategoryDetailsScreenViewModel` as they are no longer used.
- Refactored `CategoryDetailsScreen` to be a simpler composable without its own ViewModel or state management.
- Updated `CategoriesScreenUIState`:
    - Introduced a `Status` enum (Normal, Loading, NetworkError, UnknownError) to manage different UI states.
    - Changed `categories` in `CategoriesUIState` to be a list of `CategoryUiState`.
    - Added `selectedTabIndex` to `CategoriesUIState`.
    - Created `CategoryUiState` data class to hold `@StringRes name` and `@DrawableRes icon`.
- Updated `CategoriesScreenInteractionListener`:
    - Changed `onCategoryClick` parameter from `Category` to `CategoryUiState`.
    - Added `onRetry` and `onSelectTab` functions.
- Updated `CategoriesScreenViewModel` to implement the new `CategoriesScreenInteractionListener` methods.
- Revamped `CategoriesScreenContent`:
    - Added `AppTopBar` with the title "Categories".
    - Implemented a `TabRow` for "Movies" and "TV Shows".
    - Added loading placeholders and network error handling based on the `Status` in `CategoriesScreenUIState`.
    - Introduced a `CategoriesList` composable (implementation pending).
- Added new string resources for "Movies" and "TV Shows" in both English and Arabic.

Joseph-Sameh-0 committed:
Refactor: Implement CategoriesScreen UI and tab selection logic

This commit introduces the following changes:

- **CategoriesScreen.kt**:
    - Implemented the `CategoriesList` composable using `LazyVerticalGrid` to display categories.
    - Used `CategoryCard` to display individual category items.
    - Added click listeners to category items.
- **CategoriesScreenViewModel.kt**:
    - Implemented `onSelectTab` to update the selected tab index in the UI state.
    - Commented out `TODO`s for `onCategoryClick` and `onRetry`.

Joseph-Sameh-0 committed:
Refactor: Use Category enum in media use cases and repository

Updated `GetTvShowsByCategoryUseCase`, `GetMoviesByCategoryUseCase`, and `MediaRepository` to use the `Category` enum instead of integer IDs for fetching media by category.

This change also affects `MediaRepositoryImpl` to correctly map the `Category` enum to its corresponding ID when interacting with the remote data source.

ahmedNaser7 committed:
Merge branch 'develop' into feature/categories

ahmedNaser7 commented on domain/media/src/main/java/com/paris_2/domain/media/exception/AflamiException.kt:
```diff
@@ -22,6 +22,12 @@ class NoProductionCompanyFoundException(message: String = "No found production c
 class FailedToAddRatingException(message: String = "Failed to add rating") :
     AflamiException(message)
 
+class NoMoviesByCategoryException(message: String = "No Movies By Category") :
```
you can make it only one exception  for no items for specific genres

ahmedNaser7 replied:
@Joseph-Sameh-0

Joseph-Sameh-0 replied:
I didn't it 
I was working on the branch

Joseph-Sameh-0 committed:
Refactor: Update categories navigation and UI state

This commit refactors the categories feature by:

- Renaming `CategoriesDestination` and `CategoriesGraph` to `Destination` and `Graph` respectively for more generic navigation handling.
- Modifying `CategoryDetailsScreen` to accept a `category` string parameter for navigation.
- Updating `CategoriesScreenUIState` to differentiate between `moviesCategories` and `tvShowsCategories`.
- Adding `category: Category` to `CategoryUiState`.
- Removing the `onRetry` function from `CategoriesScreenInteractionListener` and the associated loading/error states from `CategoriesScreen`.
- Updating `CategoriesScreenViewModel` to navigate to `CategoryDetailsScreen` on category click, removing dependencies on `GetMoviesByCategoryUseCase` and `GetTvShowsByCategoryUseCase`.
- Adding a new string resource "action" and its Arabic translation.

Joseph-Sameh-0 committed:
Merge branch 'feature/categories' of https://github.com/Paris-Squad-S2/Aflami into feature/categories

Joseph-Sameh-0 committed:
Refactor: Update GetMoviesByCategory and GetTvShowsByCategory use cases to use Category enum

ahmedNaser7 reviewed (changes requested):
refactor the exception as i menthion in comment

Joseph-Sameh-0 committed:
Merge branch 'develop' into feature/categories

ahmedNaser7 committed:
Merge branch 'develop' into feature/categories

Joseph-Sameh-0 committed:
Populate movie categories in CategoriesScreen

Populates the `moviesCategories` list in `CategoriesScreenUIState` with various movie genres, including their names, icons, and corresponding `Category` enum values.

Also includes:
- Added string resources for new category names in English and Arabic.
- Added drawable resources for category icons.
- Added `navigationBarsPadding()` to the `CategoriesScreen` layout.

Joseph-Sameh-0 committed:
Merge branch 'develop' into feature/categories

Joseph-Sameh-0 committed:
Add TV show categories and update string resources

This commit introduces a list of categories for TV shows in `CategoriesScreenUIState.kt`.
It also updates the string resources in `strings.xml` and `strings-ar.xml` to include new category names and modify existing ones with line breaks for better display.
Additionally, new drawable assets for TV show category icons (kids, news, reality, soap, talk) have been added.

Joseph-Sameh-0 committed:
Refactor: Add MediaType to CategoriesScreenUIState

This commit updates the `CategoriesScreenUIState` and `CategoryUiState` data classes to include a `type` field of `MediaType`.

This change allows for differentiating between movie and TV show categories within the UI state. The `moviesCategories` and `tvShowsCategories` lists have been updated to correctly assign `MediaType.Movie` and `MediaType.TvShow` respectively to each `CategoryUiState`.

Joseph-Sameh-0 committed:
Refactor: Simplify Categories feature navigation

This commit refactors the navigation within the Categories feature by:

- Removing the custom NavGraph, Navigator, Destination, and related event classes.
- Introducing `CategoryDetailsActivity` to display category details.
- Navigation to `CategoryDetailsActivity` is now handled via an Intent, passing `CategoryUiState` as a JSON string.
- `CategoryUiState` is now `@Serializable` to support this.
- Removed `onCategoryClick` from `CategoriesScreenInteractionListener` as navigation is handled directly.
- Updated `CategoriesScreenViewModel` to remove navigation logic.
- Extracted `CategoriesList` into its own composable file.
- Added Hilt and Serialization plugin to the `categoriesUi` module.
- Updated `FeatureAPIModule` to remove the `CategoriesNavigator` provider.
- `CategoriesFeatureAPIImpl` now directly invokes `CategoriesScreen`.
- Added `CategoryDetailsActivity` to the AndroidManifest.

Joseph-Sameh-0 committed:
Refactor: Relocate CategoryUiState and related classes to shared package

Moved `CategoryUiState`, `Status`, `BaseViewModel`, and JSON (de)serialization functions (`toJson`, `fromJsonToCategoryUiState`) from their respective locations within the `categoriesUi` feature to a new `shared` package.

This refactoring aims to improve code organization and reusability by centralizing these common UI-related components.

Affected files:
- `CategoryDetailsActivity.kt`
- `CategoriesScreenUIState.kt`
- `navigateToMediaDetails.kt`
- `CategoryDetailsScreen.kt`
- `CategoriesScreenViewModel.kt`
- `CategoriesList.kt`

New files created:
- `feature/categories/categoriesUi/src/main/java/com/feature/categories/categoriesUi/shared/Status.kt`
- `feature/categories/categoriesUi/src/main/java/com/feature/categories/categoriesUi/shared/CategoryUiState.kt`

Renamed file:
- `feature/categories/categoriesUi/src/main/java/com/feature/categories/categoriesUi/common/BaseViewModel.kt` to `feature/categories/categoriesUi/src/main/java/com/feature/categories/categoriesUi/shared/BaseViewModel.kt`

Joseph-Sameh-0 committed:
Refactor: Implement animated visibility for categories in CategoriesScreen

Joseph-Sameh-0 committed:
Implement initial Category Details Screen

This commit introduces the Category Details screen, which displays media items for a selected category.

Key changes include:
- Added `CategoryDetailsScreen.kt` with UI for displaying category details, including a back button and a list of categories.
- Implemented `CategoryDetailsScreenViewModel.kt` to manage UI state and handle category selection.
- Created `CategoryDetailsScreenUIState.kt` to define the screen's UI state.
- Added `CategoryDetailsScreenInteractionListener.kt` for handling user interactions.
- Introduced an `Empty` state to `Status.kt` for handling cases where no items are found for a genre.
- Added a new string resource `no_items_for_this_genre` and its Arabic translation for the empty state.
- Included new drawable resources for displaying an image in the empty state.
- Updated `CategoryDetailsScreen.kt` to show a placeholder view when the `Status` is `Empty`.
- The screen now initializes with the provided category and dynamically updates the title and category list based on the media type (Movie/TV Show).
- Media items are displayed with an animated visibility transition.

Joseph-Sameh-0 committed:
Implement initial Category Details Screen

This commit introduces the Category Details screen, which displays media items for a selected category.

Key changes include:
- Added `CategoryDetailsScreen.kt` with UI for displaying category details, including a back button and a list of categories.
- Implemented `CategoryDetailsScreenViewModel.kt` to manage UI state and handle category selection.
- Created `CategoryDetailsScreenUIState.kt` to define the screen's UI state.
- Added `CategoryDetailsScreenInteractionListener.kt` for handling user interactions.
- Introduced an `Empty` state to `Status.kt` for handling cases where no items are found for a genre.
- Added a new string resource `no_items_for_this_genre` and its Arabic translation for the empty state.
- Included new drawable resources for displaying an image in the empty state.
- Updated `CategoryDetailsScreen.kt` to show a placeholder view when the `Status` is `Empty`.
- The screen now initializes with the provided category and dynamically updates the title and category list based on the media type (Movie/TV Show).
- Media items are displayed with an animated visibility transition.

Joseph-Sameh-0 committed:
Implement MediaCard component and integrate into CategoryDetailsScreen

This commit introduces a new `MediaCard` composable for displaying media items (movies/TV shows) with details like poster, rating, title, type, and year.

Key changes:
- Added `MediaCard.kt` with the new composable and its previews.
- Added `RatingCard.kt` as a sub-component for `MediaCard` to display ratings.
- Added `DescriptionSeparator.kt` to format media type and year within `MediaCard`.
- Created `MediaUI` data class to represent media items in the UI layer, replacing direct usage of the domain `Media` entity in `CategoryDetailsScreenViewModel` and `CategoryDetailsScreenInteractionListener`.
- Integrated `MediaCard` into `CategoryDetailsMediaList.kt` to display a grid of media items.
- Updated `CategoryDetailsScreen.kt` to use `navigationBarsPadding()` correctly.
- Added new string resources for "Movie" and "TV show".
- Included `safeimageviewer` and `datetime` dependencies in `build.gradle.kts`.

Joseph-Sameh-0 committed:
Refactor: Replace NetworkError with CategoryDetailsEmptyScreen in CategoryDetailsScreen

Updated `CategoryDetailsScreen.kt` to display `CategoryDetailsEmptyScreen` instead of `NetworkError` when an unknown error occurs.

Joseph-Sameh-0 committed:
Refactor: Standardize exception handling in repositories

Replaced specific exception types (e.g., `NoTvShowFoundException`, `NoMovieFoundException`, `NoCategoriesFoundException`) with a generic `FailedException` across various repositories. This change simplifies error handling by providing a consistent exception type for different failure scenarios. The specific context of the failure is now passed as a message to the `FailedException` constructor.

Affected repositories include:
- `TvShowRepositoryImpl`
- `MovieRepositoryImpl`
- `SearchMediaRepositoryImpl`
- `CategoriesRepositoryImpl`
- `MoviesCategoriesRepositoryImpl`
- `MediaRepositoryImpl`

Additionally, made DTO properties in `ResultDto` and `MovieByCategoryDto` nullable to prevent potential `NullPointerExceptions` during deserialization.
Updated `MediaMapper` to handle nullable fields in `ResultDto` when converting to the domain `Media` object.

Joseph-Sameh-0 committed:
Refactor: Simplify AflamiException hierarchy

Removed several unused custom exception classes and updated `NoInternetConnectionException` to use a default message.

Joseph-Sameh-0 committed:
Refactor: Standardize exception handling in repository tests

Replaced specific custom exceptions (e.g., `NoMediaForActorException`, `NoCategoriesFoundException`, `NoMovieFoundException`, `NoCastFoundException`, `NoGalleryFoundException`, `FailedToDeleteRatingException`, `AddMediaToContinueWatchingException`, `GetContinueWatchingMediaException`, `NoRatedMediaFoundException`) with a generic `FailedException` across multiple repository tests.

This change affects the following test files:
- `SearchMediaRepositoryImplTest.kt`
- `MoviesCategoriesRepositoryImplTest.kt`
- `CategoriesRepositoryImplTest.kt`
- `MovieRepositoryImplTest.kt`
- `MediaRepositoryImplTest.kt`

Joseph-Sameh-0 committed:
Merge branch 'develop' into feature/categories

Joseph-Sameh-0 committed:
Refactor: Integrate Paging 3 for media list in CategoryDetails

This commit refactors the `CategoryDetails` feature to use the Paging 3 library for loading and displaying media lists.

Key changes:
- Introduced a generic `PagingSource` to handle paginated data loading for media items.
- Updated `CategoryDetailsScreenViewModel` to use `Pager` and the new `PagingSource` to provide a `Flow<PagingData<MediaUI>>`.
- Modified `CategoryDetailsMediaList` composable to accept `LazyPagingItems<MediaUI>` and display paginated content.
- Updated `CategoryDetailsScreen` to collect `LazyPagingItems` and handle Paging 3 `LoadState` for loading, error, and empty states.
- Removed the custom `tryToExecute` logic and manual status handling (`Status` enum) from `BaseViewModel` and `CategoryDetailsScreenViewModel` as Paging 3 provides built-in mechanisms for this.
- Added Paging 3 library dependency to `feature/categories/categoriesUi/build.gradle.kts`.
- Made `CategoriesScreenContent` and `CategoriesScreenContent` (in `CategoryDetailsScreen`) private as they are not used outside their respective files.

Joseph-Sameh-0 committed:
Merge branch 'develop' into feature/categories

Joseph-Sameh-0 committed:
Apply saved app language on CategoryDetailsActivity

Introduced `MainViewModel` and `InstallSavedAppLanguage` composable to retrieve and apply the user's last selected language preference when `CategoryDetailsActivity` is created.

- Added `projects.domain.user` dependency to `feature/categories/categoriesUi/build.gradle.kts`.
- Created `MainViewModel.kt` to expose the `getLastSelectedAppLanguage` function from `SettingsUseCase`.
- Created `InstallSavedAppLanguage.kt` which observes the selected language and updates the app's locale.
- Updated `CategoryDetailsActivity.kt`:
    - Changed parent class from `ComponentActivity` to `AppCompatActivity`.
    - Injected `SettingsUseCase`.
    - Called `InstallSavedAppLanguage` within `setContent`.
    - Used `settingsUseCase.isDarkTheme()` to determine the theme for `AflamiTheme`.

Joseph-Sameh-0 committed:
Refactor: Standardize exception handling in TvShowRepositoryImplTest

Updated TvShowRepositoryImplTest to use a generic `FailedException` for various error scenarios, replacing more specific exceptions like `FailedToAddRatingException`, `NoCastFoundException`, `NoGalleryFoundException`, `NoSeasonFoundException`, `NoTvShowFoundException`, and `FailedToDeleteRatingException`.

MustafaIbrahim96 reviewed (commented):
Greate work , keep going

Joseph-Sameh-0 committed:
Refactor: Update Hilt view models to use javax.inject.Inject

This commit updates the Hilt view models in the categories feature to use `javax.inject.Inject` instead of `jakarta.inject.Inject`. This change affects `MainViewModel`, `CategoryDetailsScreenViewModel`, and `CategoriesScreenViewModel`.
